[all hosts] (FRE) Add tutorial (#4716)
OfficeDev/office-js-docs-pr: docs/tutorials/first-run-experience-tutorial.md

        // showedFRE is created and set to "true" when you call showFirstRunExperience().
        if (!localStorage.getItem("showedFRE")) {
          showFirstRunExperience();
        }
    
        document.getElementById("run").onclick = run;
      }
    });
    
    async function showFirstRunExperience() {
      document.getElementById("first-run-experience").style.display = "flex";
      localStorage.setItem("showedFRE", true);
    }  
    ```

1. If you selected an Office application besides Excel, update the condition of the first `if` statement to check for your chosen [Office.HostType](/javascript/api/office/office.hosttype).

### Update the CSS file

Update the CSS file to ensure that the add-in UI is styled appropriately given the addition of the "first-run-experience" `<div>`.

- Open the **taskpane.css** file. Replace the line `.ms-welcome__main {` with the following code, then save the file.

    ```css
    .ms-welcome__main, .ms-welcome__main > div {
    ```

### Try it out

1. Ensure that the web server is running and the add-in has been sideloaded. For details, see the instructions in the quick start you used.

> 1. If the add-in task pane isn't already open in Office, go to the **Home** tab and choose the **Show Taskpane** button on the ribbon to open it.

Reviewer: Consider adding a note here for Outlook since the location of the add-in varies depending on the client. Maybe we can refer the user back to the quick start or to [Use add-ins in Outlook](https://support.microsoft.com/office/1ee261f9-49bf-4ba6-b3e2-2ba7bcab64c8) for instructions.

---
title: First-run experience tutorial
description: Learn how to implement a first-run experience for your Office Add-in.
ms.date: 08/09/2024
ms.topic: how-to
ms.localizationpriority: medium
---

# Build an Office Add-in with a basic first-run experience

> In this article, you'll walk through the process of updating a task pane add-in to include a [first-run experience](../design/first-run-experience-patterns.md) using the value placemat pattern. When the user runs the add-in, the add-in determines whether or not to show the first-run experience by checking [local storage](../develop/persisting-add-in-state-and-settings.md#browser-storage) for the existence of a certain flag.

Reviewer 2: Suggestion to make this less wordy.
```suggestion
In this article, you'll walk through the process of updating a task pane add-in to include a [first-run experience](../design/first-run-experience-patterns.md) using the value placemat pattern. When the user runs the add-in, the add-in determines whether or not to show the first-run experience by checking [local storage](../develop/persisting-add-in-state-and-settings.md#browser-storage) for a flag.
```

> This tutorial provides instructions and screenshots for Excel but you can use a similar pattern to implement a first-run experience in other Office applications where Office Web Add-ins are supported. Also note that actual interactions with your selected Office application will differ.

Reviewer 2: > Also note that actual interactions with your selected Office application will differ.
I'm not sure what this means in this context. Is the code different or the UI flow?

Author: @AlexJerabek We add an instruction for the user to select cells before clicking Run. I was thinking about a situation like that where it's only applicable to Excel. That said, I'll remove that sentence since I have a step that asks the dev to update the instruction

## Prerequisites

1. Select the quick start for the Office application you'd like to use. Make sure to choose the Yeoman generator version.

    - [Excel](../quickstarts/excel-quickstart-jquery.md)
    - [OneNote](../quickstarts/onenote-quickstart.md)
    - [Outlook](../quickstarts/outlook-quickstart.md)
    - [PowerPoint](../quickstarts/powerpoint-quickstart.md)
    - [Project](../quickstarts/project-quickstart.md)
    - [Word](../quickstarts/word-quickstart.md)

1. Follow the instructions in your selected quick start. After you complete its "Try it out" section, return here to continue.

## Implement the first-run experience

### Update the HTML file

> Be clear about the area of the UI that will be part of the first-run experience. In this tutorial, you'll create a div with `id` named "first-run-experience" that represents what users see only the first time they run your add-in.

Reviewer 2: Here and elsewhere, should div be in code ticks and angle brackets?

> 1. Open the **taskpane.html**. Replace the `<main>` element with the following markup, then save the file. Note about this markup:

Reviewer 2: Since there is more than one note, consider pluralizing the lead-in.
```suggestion
1. Open the **taskpane.html**. Replace the `<main>` element with the following markup, then save the file. Some notes about this markup:
```

Author: @AlexJerabek We may need to clean this up because I got this terminology from elsewhere in the docs

                    <span class="ms-font-m">Achieve more with Office integration</span>
                </li>
                <li class="ms-ListItem">
                    <i class="ms-Icon ms-Icon--Unlock ms-font-xl"></i>
                    <span class="ms-font-m">Unlock features and functionality</span>
                </li>
                <li class="ms-ListItem">
                    <i class="ms-Icon ms-Icon--Design ms-font-xl"></i>
                    <span class="ms-font-m">Create and visualize like a pro</span>
                </li>
            </ul>
        </div>
        <p class="ms-font-l">Select any range of cells in the worksheet, then click <b>Run</b>.</p>
        <div role="button" id="run" class="ms-welcome__action ms-Button ms-Button--hero ms-font-xl">
            <span class="ms-Button-label">Run</span>
        </div>
        <p><label id="item-subject"></label></p>    
    </main>
    ```

1. If you selected an Office application besides Excel, update the first `<p>` element with more appropriate instructions.

### Update the JavaScript file

Update the JavaScript file to display the first-run experience if this is the first time the user is running the add-in.

> 1. Open the **taskpane.js** file. Replace the `Office.onReady` statement with the following code, then save the file. Note about this code:

Reviewer 2: ```suggestion
1. Open the **taskpane.js** file. Replace the `Office.onReady` statement with the following code, then save the file. Some notes about this code:
```

>     - It checks local storage for a key called "showedFRE". If it doesn't exist, then show the first-run experience.

Reviewer 2: Since "it" refers to the add-in in the previous sentence, consider clarifying the noun in the second sentence.
```suggestion
    - It checks local storage for a key called "showedFRE". If the key doesn't exist, then show the first-run experience.
```

>     - Adds a new function called `showFirstRunExperience` that displays the "first-run-experience" div added to the HTML. This function also adds the "showedFRE" item to local storage.

Reviewer 2: To be consistent with the previous bullet.
```suggestion
    - It adds a new function called `showFirstRunExperience` that displays the "first-run-experience" div added to the HTML. This function also adds the "showedFRE" item to local storage.
```

    ```javascript
    Office.onReady((info) => {
      if (info.host === Office.HostType.Excel) {
        document.getElementById("sideload-msg").style.display = "none";
        document.getElementById("app-body").style.display = "flex";
    

>         if (!localStorage.getItem("showedFRE")) {

Reviewer 2: Consider adding a clarifying comment that the setting is created after the first run and won't ever be false.

Author: @AlexJerabek Added

    ```

### Try it out

1. Ensure that the web server is running and the add-in has been sideloaded. For details, see the instructions in the quick start you used.

1. If the add-in task pane isn't already open in Office, go to the **Home** tab and choose the **Show Taskpane** button on the ribbon to open it.

1. Verify that the task pane includes the list of features.

    ![The add-in task pane UI on first run.](../images/fre-tutorial-addin-first-run.png)

1. Close the task pane then reopen it. Verify that the task pane no longer displays the list of features.

    ![The add-in task pane UI on second run.](../images/fre-tutorial-addin-next-run.png)

1. If you used `npm start` to start the local web server, run the following command when you want to stop the server and uninstall the add-in.

    ```command&nbsp;line
    npm stop
    ```

## Next steps

Congratulations, you've successfully created an Office task pane add-in with a first-run experience!

### Make it production ready

Using this tutorial, you implemented a basic [first-run experience](../design/first-run-experience-patterns.md). For the first-run experience to be ready for users, you should consider the following:

> - Use a more secure and robust option for tracking first-run state. For example, use storage partitioning if available, or implement a Single Sign-on (SSO) authentication solution. For more about available settings options, see [Persist add-in state and settings](../develop/persisting-add-in-state-and-settings.md). For more about available authentication options, see [Overview of authentication and authorization](../develop/overview-authn-authz.md).

Reviewer 2: Consider moving this bullet to the last one, since it's the least actionable in most cases.

- Update the features listed in the value placemat to match what your add-in actually does.
- Implement a different pattern (for example, video placemat or carousel) that better showcases the benefits of your add-in.

If you're planning to make your add-in available in the AppSource marketplace, you'll need to have a robust and useful first-run experience. For more information, see [Best practices for developing Office Add-ins](../concepts/add-in-development-best-practices.md).

## Code samples

> - ["Hello world" add-ins](../overview/office-add-in-code-samples.md#getting-started): Learn how to build a simple Office Add-in with only a manifest, HTML web page, and a logo.

Reviewer 2: Since this sample doesn't have an FRE, I'm not sure it makes sense to link to it. Maybe we should add a completed FRE tutorial sample and link to that?

Author: I like the idea of a completed tutorial. I'll add that as a work item. In the meantime, I'll remove this section and maybe add the link to the "See also" section
